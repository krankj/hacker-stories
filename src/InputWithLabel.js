import React from "react";
import styles from "./App.module.css";
const InputWithLabel = ({
  id,
  children,
  type = "text",
  value,
  isFocused,
  onInputChange,
}) => {
  const inputRef = React.useRef();
  React.useEffect(() => {
    if (isFocused) {
      inputRef.current.focus();
    }
  }, [isFocused]);
  return (
    <>
      <label htmlFor={id} className={styles.label}>
        {children}
      </label>
      &nbsp;
      <input
        ref={inputRef}
        id={id}
        type={type}
        value={value}
        onChange={onInputChange}
        className={styles.input}
      />
    </>
  );
};

export default InputWithLabel;
