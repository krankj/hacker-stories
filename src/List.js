import React from "react";
import styles from "./App.module.css";
import { ReactComponent as Dismiss } from "./close.svg";
import SortButton from "./SortButton";
import { sortBy } from "lodash";

const List = React.memo(({ list, onRemoveItem }) => {
  const [sort, setSort] = React.useState(
    JSON.parse(localStorage.getItem("sortInfo")) || {
      sortOnColumn: "NONE",
      sortAsc: true,
    }
  );

  React.useEffect(() => {
    localStorage.setItem("sortInfo", JSON.stringify(sort));
  }, [sort]);

  const SORT = {
    NONE: (list) => list,
    TITLE: (list) => sortBy(list, "title"),
    AUTHOR: (list) => sortBy(list, "author"),
    COMMENTS: (list) => sortBy(list, "num_comments"),
    POINTS: (list) => sortBy(list, "points"),
  };

  const sortFunction = SORT[sort.sortOnColumn];

  const sortedList = sort.sortAsc
    ? sortFunction(list)
    : sortFunction(list).reverse();

  const handleSort = (sortOnColumn) => {
    setSort((prevState) =>
      prevState.sortOnColumn !== sortOnColumn
        ? { ...prevState, sortOnColumn }
        : { ...prevState, sortAsc: !prevState.sortAsc }
    );
  };

  const headers = [
    {
      id: 1,
      title: "Title",
      width: "40%",
    },
    {
      id: 2,
      title: "Author",
      width: "30%",
    },
    {
      id: 3,
      title: "Comments",
      width: "10%",
    },
    {
      id: 4,
      title: "Points",
      width: "10%",
    },
  ];

  headers.map((header) => {
    if (sort.sortOnColumn === header.title.toUpperCase()) {
      header.visible = true;
    } else header.visible = false;
    return header;
  });

  return (
    <>
      <div className={styles.sortHeader}>
        {headers.map((header) => {
          return (
            <span style={{ width: header.width }}>
              <SortButton
                sortAsc={sort.sortAsc}
                key={header.id}
                visible={header.visible}
                onSortHandler={handleSort}
              >
                {header.title}
              </SortButton>
            </span>
          );
        })}
      </div>
      {sortedList.map((item) => (
        <Item key={item.objectID} item={item} onRemoveItemLast={onRemoveItem} />
      ))}
    </>
  );
});
const Item = ({ item, onRemoveItemLast }) => (
  <div className={styles.item}>
    <span style={{ width: "40%" }}>
      <a href={item.url}>{item.title}</a>
    </span>

    <span style={{ width: "30%" }}>{item.author}</span>

    <span style={{ width: "10%" }}>{item.num_comments}</span>

    <span style={{ width: "10%" }}>{item.points}</span>

    <span style={{ width: "10%" }}>
      <button
        className={`${styles.button} ${styles.buttonSmall}`}
        type="button"
        onClick={() => onRemoveItemLast(item)}
      >
        <Dismiss height="14px" width="14px" />
      </button>
    </span>
  </div>
);

export { List, Item };
