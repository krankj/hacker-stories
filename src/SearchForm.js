import React from "react";
import styles from "./App.module.css";
import InputWithLabel from "./InputWithLabel";

const SearchForm = ({ searchTerm, onSearchInput, onSubmit }) => (
  <form onSubmit={onSubmit} className={styles.searchForm}>
    <InputWithLabel
      id="search"
      onInputChange={onSearchInput}
      isFocused
      value={searchTerm}
    >
      Search:
    </InputWithLabel>
    <button
      className={`${styles.button} ${styles.buttonSmall}`}
      type="submit"
      disabled={!searchTerm}
    >
      Submit
    </button>
  </form>
);
export default SearchForm;
