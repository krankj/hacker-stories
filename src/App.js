import React from "react";
import axios from "axios";
import styles from "./App.module.css";
import { List } from "./List.js";
import SearchForm from "./SearchForm";
import LastSearchButton from "./LastSearchButton";
import InputWithLabel from "./InputWithLabel";
// type Story = {
//   objectID: string;
//   url: string;
//   title: string;
//   author: string;
//   num_comments: number;
//   points: number;
// };

// type SortObj = {
//   sortOnColumn: string;
//   sortAsc: boolean;
// };

// type InputWithLabelProps = {
//   id: string;
//   value: string;
//   type?: string;
//   onInputChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
//   isFocused?: boolean;
//   children: React.ReactNode;
// };

// type StoriesState = {
//   data: Stories;
//   isLoading: boolean;
//   isError: boolean;
// };
// type Stories = Array<Story>;

// type ListProps = {
//   list: Stories;
//   onRemoveItem: (item: Story) => void;
// };

// type SearchFormProps = {
//   searchTerm: string;
//   onSearchInput: (event: React.ChangeEvent<HTMLInputElement>) => void;
//   onSubmit: (event: React.FormEvent<HTMLFormElement>) => void;
// };

// interface StoriesFetchInitAction {
//   type: "STORIES_FETCH_INIT";
// }
// interface StoriesFetchSuccessAction {
//   type: "STORIES_FETCH_SUCCESS";
//   payload: Stories;
// }
// interface StoriesFetchFailureAction {
//   type: "STORIES_FETCH_FAILURE";
// }
// interface StoriesRemoveAction {
//   type: "REMOVE_STORY";
//   payload: Story;
// }

// interface StoriesSortAction {
//   type: "SORT_STORIES";
//   payload: SortObj;
// }

// type StoriesAction =
//   | StoriesFetchInitAction
//   | StoriesFetchSuccessAction
//   | StoriesFetchFailureAction
//   | StoriesRemoveAction
//   | StoriesSortAction;
// type ItemProps = {
//   item: Story,
//   onRemoveItemLast: (item: Story) => void,
// };

const storiesReducer = (state, action) => {
  switch (action.type) {
    case "STORIES_FETCH_SUCCESS":
      let concatedResult = [...state.data, ...action.payload.data];
      return {
        ...state,
        data: action.payload.page === 0 ? action.payload.data : concatedResult,
        isLoading: false,
      };
    case "REMOVE_STORY":
      return {
        ...state,
        data: state.data.filter(
          (item) => action.payload.objectID !== item.objectID
        ),
      };
    case "STORIES_FETCH_INIT":
      return { ...state, isLoading: true, isError: false };
    case "STORIES_FETCH_FAILURE":
      return { ...state, isLoading: false, isError: true };

    default:
      throw new Error("Something went wrong");
  }
};
const API_ENDPOINT = "https://hn.algolia.com/api/v1/search?query=";

const App = () => {
  const useSemiPersistentState = (key, initialState) => {
    const [value, setValue] = React.useState(
      localStorage.getItem(key) || initialState
    );

    const isMounted = React.useRef(false);
    React.useEffect(() => {
      if (isMounted.current) {
        localStorage.setItem(key, value);
      } else {
        isMounted.current = true;
      }
    }, [value, key]);

    return [value, setValue];
  };

  const [searchTerm, setSearchTerm] = useSemiPersistentState("search", "React");

  const [lastSearchedTerm, setLastSearchedTerm] = React.useState("");
  const [searchTermCache, setSearchTermCache] = React.useState([]);
  const [titleSearch, setTitleSearch] = React.useState("");
  const [page, setPage] = React.useState(0);
  const [url, setUrl] = React.useState(
    `${API_ENDPOINT}${searchTerm}&page=${page}`
  );

  const [stories, dispatchStories] = React.useReducer(storiesReducer, {
    data: [],
    isLoading: false,
    isError: false,
  });

  async function testCallBack() {
    dispatchStories({ type: "STORIES_FETCH_INIT" });
    try {
      const result = await axios.get(`${url}&page=${page}`);
      dispatchStories({
        type: "STORIES_FETCH_SUCCESS",
        payload: { data: result.data.hits, page: page },
      });
      setLastSearchedTerm(searchTerm);
    } catch {
      dispatchStories({ type: "STORIES_FETCH_FAILURE" });
    }
  }

  const handleFetchStories = React.useCallback(testCallBack, [page, url]);

  React.useEffect(() => {
    handleFetchStories();
  }, [handleFetchStories]);

  const handleTitleSearch = (event) => {
    setTitleSearch(event.target.value);
  };

  const handleSubmit = (event) => {
    if (lastSearchedTerm !== searchTerm) {
      setSearchTermCache((prevSearch) => {
        let newSearch = [...prevSearch];
        if (newSearch.length >= 5) {
          newSearch.shift();
          newSearch.splice(4, 1, lastSearchedTerm);
        } else {
          newSearch = [...prevSearch, lastSearchedTerm];
        }
        return newSearch;
      });
    }
    setPage(0);
    setUrl(`${API_ENDPOINT}${searchTerm}`);
    event.preventDefault();
  };

  // const searchedList = () =>
  //   stories.data.filter((story) => {
  //     console.log("Running filter");
  //     return story.title.toLowerCase().includes(searchTerm.toLowerCase());
  //   });

  const handleRemoveStory = React.useCallback((item) => {
    navigator.vibrate =
      navigator.vibrate ||
      navigator.webkitVibrate ||
      navigator.mozVibrate ||
      navigator.msVibrate;

    if (navigator.vibrate) {
      navigator.vibrate(1000);
    }
    dispatchStories({
      type: "REMOVE_STORY",
      payload: item,
    });
  }, []);

  const handleSearchInput = (event) => {
    setSearchTerm(event.target.value);
  };

  const handleBackButton = () => {
    let prevValue = searchTermCache.pop();
    setSearchTerm(prevValue);
    setUrl(`${API_ENDPOINT}${prevValue}`);
  };

  const [isFetching, setIsFetching] = React.useState(false);

  const handleScroll = () => {
    if (
      window.innerHeight + document.documentElement.scrollTop !==
      document.documentElement.scrollHeight
    )
      return;
    setIsFetching(true);
  };

  function fetchItems() {
    setPage((prevPage) => prevPage + 1);
    setIsFetching(false);
  }

  React.useEffect(() => {
    if (!isFetching) return;
    fetchItems();
  }, [isFetching]);

  React.useEffect(() => {
    window.addEventListener("scroll", handleScroll);
    return () => window.removeEventListener("scroll", handleScroll);
  }, []);

  // Or React.useCallBack(getSumComments(stories), [stories]);
  //const sumComments = React.useMemo(() => getSumComments(stories), [stories]);
  return (
    <div className={styles.rootContainer}>
      <div className={styles.container}>
        <h1 className={styles.headlinePrimary}>My Hacker Stories</h1>
        <SearchForm
          searchTerm={searchTerm}
          onSearchInput={handleSearchInput}
          onSubmit={handleSubmit}
        />
        <LastSearchButton
          enabled={searchTermCache.length !== 0}
          onBackButtonClick={handleBackButton}
        />
        <div className={styles.titleSearchBar}>
          <InputWithLabel
            onInputChange={handleTitleSearch}
            id="titleSearch"
            isFocused={false}
            value={titleSearch}
          >
            Title:
          </InputWithLabel>
        </div>
        {stories.isLoading ? (
          <p>Loading...</p>
        ) : stories.isError ? (
          <p> Error while fetching data :( </p>
        ) : (
          <>
            <List list={stories.data} onRemoveItem={handleRemoveStory} />
            {isFetching && "Fetching more items..."}
          </>
        )}
      </div>
    </div>
  );
};

// const initialStories = [
//   {
//     title: "React",
//     url: "https://reactjs.org/",
//     author: "Jordan Walke",
//     num_comments: 3,
//     points: 4,
//     objectID: 0,
//   },
//   {
//     title: "Redux",
//     url: "https://redux.js.org/",
//     author: "Dan Abramov, Andrew Clark",
//     num_comments: 2,
//     points: 5,
//     objectID: 1,
//   },
//   {
//     title: "Real",
//     url: "https://reactjs.org/",
//     author: "Jordan Walke",
//     num_comments: 3,
//     points: 4,
//     objectID: 2,
//   },
//   {
//     title: "Red",
//     url: "https://redux.js.org/",
//     author: "Dan Abramov, Andrew Clark",
//     num_comments: 2,
//     points: 5,
//     objectID: 3,
//   },
//   {
//     title: "Rent",
//     url: "https://reactjs.org/",
//     author: "Jordan Walke",
//     num_comments: 3,
//     points: 4,
//     objectID: 4,
//   },
//   {
//     title: "Rexona",
//     url: "https://redux.js.org/",
//     author: "Dan Abramov, Andrew Clark",
//     num_comments: 2,
//     points: 5,
//     objectID: 5,
//   },
//   {
//     title: "Redeem",
//     url: "https://reactjs.org/",
//     author: "Jordan Walke",
//     num_comments: 3,
//     points: 4,
//     objectID: 6,
//   },
//   {
//     title: "Redemption",
//     url: "https://redux.js.org/",
//     author: "Dan Abramov, Andrew Clark",
//     num_comments: 2,
//     points: 5,
//     objectID: 7,
//   },
// ];

// const getAsyncStories = () =>
//   new Promise((resolve, error) => {
//     setTimeout(() => resolve({ data: { stories: initialStories } }), 2000);
//   });

export { App as default, storiesReducer };
