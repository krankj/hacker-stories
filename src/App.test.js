import React from "react";
import { render, screen, fireEvent, act } from "@testing-library/react";

import App, { storiesReducer } from "./App";

import { List, Item } from "./List.js";
import InputWithLabel from "./InputWithLabel.js";
import SearchForm from "./SearchForm.js";

import axios from "axios";
jest.mock("axios");

const storyOne = {
  title: "React",
  url: "https://reactjs.org/",
  author: "Jordan Walke",
  num_comments: 3,
  points: 4,
  objectID: 0,
};

const storyTwo = {
  title: "Redux",
  url: "https://redux.js.org/",
  author: "Dan Abramov, Andrew Clark",
  num_comments: 2,
  points: 5,
  objectID: 1,
};

const stories = [storyOne, storyTwo];

describe("Component testing", () => {
  test("Render Item component", () => {
    render(<Item item={storyOne} />);
    expect(screen.getByText("Jordan Walke")).toBeInTheDocument();
    expect(screen.getByText("React")).toHaveAttribute(
      "href",
      "https://reactjs.org/"
    );
  });
  test("Make sure there is a dismiss button", () => {
    render(<Item item={storyOne} />);
    expect(screen.getByRole("button")).toBeInTheDocument();
  });
  test("Clicking dismiss causes the callback handler", () => {
    const onDismissHandler = jest.fn();
    render(<Item item={storyOne} onRemoveItemLast={onDismissHandler} />);
    fireEvent.click(screen.getByRole("button"));
    expect(onDismissHandler).toHaveBeenCalledTimes(1);
  });
});

describe("App component", () => {
  test("Data fetch success", async () => {
    const promise = Promise.resolve({
      data: {
        hits: stories,
      },
    });

    axios.get.mockImplementationOnce(() => promise);
    render(<App />);
    expect(screen.queryByText(/Loading/)).toBeInTheDocument();
    await act(() => promise);
    expect(screen.queryByText(/Loading/)).toBe(null);
    expect(screen.getAllByText("React")[0]).toBeInTheDocument();
    expect(screen.getByText("Redux")).toBeInTheDocument();
    expect(screen.getAllByText("check.svg").length).toBe(2);
  });

  test("Data fetch failure", async () => {
    const promise = Promise.reject();
    axios.get.mockImplementationOnce(() => promise);
    render(<App />);
    expect(screen.queryByText(/Loading/)).toBeInTheDocument();
    try {
      await act(() => promise);
    } catch {
      expect(screen.queryByText(/Loading/)).not.toBeInTheDocument();
      expect(screen.queryByText(/Error/)).toBeInTheDocument();
    }
  });

  test("Removes a story from the app component", async () => {
    const promise = Promise.resolve({
      data: {
        hits: stories,
      },
    });

    axios.get.mockImplementationOnce(() => promise);
    render(<App />);
    await act(() => promise);
    expect(screen.getAllByText("check.svg").length).toBe(2);
    expect(screen.getByText("Jordan Walke")).toBeInTheDocument();
    fireEvent.click(screen.getAllByText("check.svg")[0]);
    expect(screen.getAllByText("check.svg").length).toBe(1);
    expect(screen.queryByText("Jordan Walke")).not.toBeInTheDocument();
  });

  test("Search for a component and render a new list", async () => {
    const reactPromise = Promise.resolve({
      data: {
        hits: stories,
      },
    });

    const anotherStory = {
      title: "JavaScript",
      url: "https://en.wikipedia.org/wiki/JavaScript",
      author: "Brendan Eich",
      num_comments: 15,
      points: 10,
      objectID: 3,
    };

    const javascriptPromise = Promise.resolve({
      data: {
        hits: [anotherStory],
      },
    });

    axios.get.mockImplementation((url) =>
      url.includes("React")
        ? reactPromise
        : url.includes("JavaScript")
        ? javascriptPromise
        : () => {
            throw Error();
          }
    );

    render(<App />);

    await act(() => reactPromise);

    expect(screen.queryByDisplayValue("React")).toBeInTheDocument();
    expect(screen.queryByDisplayValue("JavaScript")).toBeNull();

    expect(screen.queryByText("Jordan Walke")).toBeInTheDocument();
    expect(screen.queryByText("Dan Abramov, Andrew Clark")).toBeInTheDocument();
    expect(screen.queryByText("Brendan Eich")).toBeNull();

    fireEvent.change(screen.queryByDisplayValue("React"), {
      target: {
        value: "JavaScript",
      },
    });

    expect(screen.queryByDisplayValue("JavaScript")).toBeInTheDocument();
    expect(screen.queryByDisplayValue("React")).toBeNull();

    fireEvent.submit(screen.queryByText("Submit"));

    await act(() => javascriptPromise);
    expect(screen.queryByText("Jordan Walke")).toBeNull();
    expect(screen.queryByText("Dan Abramov, Andrew Clark")).toBeNull();
    expect(screen.queryByText("Brendan Eich")).toBeInTheDocument();
  });

  test("App snapshot component", async () => {
    const { container } = render(<App />);
    const promise = Promise.resolve({
      data: {
        hits: stories,
      },
    });
    await act(() => promise);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Search Component ", () => {
  const searchFromProps = {
    searchTerm: "React",
    onSubmit: jest.fn(),
    onSearchInput: jest.fn(),
  };
  test("Render search component", () => {
    render(<SearchForm {...searchFromProps} />);

    expect(screen.getByDisplayValue("React")).toBeInTheDocument();
  });
  test("Renders the correct label", () => {
    render(<SearchForm {...searchFromProps} />);
    expect(screen.getByLabelText(/Search/)).toBeInTheDocument();
  });

  test("Trigger onSearchInput for a field update", () => {
    render(<SearchForm {...searchFromProps} />);
    fireEvent.change(screen.getByDisplayValue("React"), {
      target: { value: "Redux" },
    });
    expect(searchFromProps.onSearchInput).toHaveBeenCalledTimes(1);
  });

  test("Trigger onSubmit when the button is clicked", () => {
    render(<SearchForm {...searchFromProps} />);
    fireEvent.submit(screen.getByRole("button"));
    expect(searchFromProps.onSubmit).toHaveBeenCalledTimes(1);
  });

  test("Snapshot test", () => {
    const { container } = render(<SearchForm {...searchFromProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Stories Reducer", () => {
  test("Removes a story from the story list", () => {
    const state = { data: stories, isLoading: false, isError: false };
    const action = { type: "REMOVE_STORY", payload: storyTwo };
    const newState = storiesReducer(state, action);
    const expectedState = {
      data: [storyOne],
      isLoading: false,
      isError: false,
    };
    expect(newState).toStrictEqual(expectedState);
  });

  test("Fetch init", () => {
    const state = { data: stories, isLoading: false, isError: false };
    const action = { type: "STORIES_FETCH_INIT" };
    const newState = storiesReducer(state, action);

    const expectedState = {
      data: stories,
      isLoading: true,
      isError: false,
    };

    expect(newState).toStrictEqual(expectedState);
  });

  test("Fetch failed", () => {
    const state = { data: stories, isLoading: false, isError: false };
    const action = { type: "STORIES_FETCH_FAILURE" };
    const newState = storiesReducer(state, action);
    const expectedState = { data: stories, isLoading: false, isError: true };
    expect(newState).toStrictEqual(expectedState);
  });

  test("Fetch Success", () => {
    const state = { data: [], isLoading: false, isError: false };
    const action = { type: "STORIES_FETCH_SUCCESS", payload: stories };
    const newState = storiesReducer(state, action);
    const expectedState = { data: stories, isLoading: false, isError: false };
    expect(newState).toStrictEqual(expectedState);
  });
});
