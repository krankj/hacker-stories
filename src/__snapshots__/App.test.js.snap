// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component App snapshot component 1`] = `
<div
  class="container"
>
  <h1
    class="headlinePrimary"
  >
    My Hacker Stories
  </h1>
  <form
    class="searchForm"
  >
    <label
      class="label"
      for="search"
    >
      Search:
    </label>
     
    <input
      class="input"
      id="search"
      type="text"
      value="JavaScript"
    />
    <p>
      Searching : 
      <strong>
        JavaScript
      </strong>
    </p>
    <button
      class="button buttonSmall"
      type="submit"
    >
      Submit
    </button>
  </form>
  <div
    class="item"
  >
    <span
      style="width: 40%;"
    >
      <a
        href="https://en.wikipedia.org/wiki/JavaScript"
      >
        JavaScript
      </a>
    </span>
    <span
      style="width: 30%;"
    >
      Brendan Eich
    </span>
    <span
      style="width: 10%;"
    >
      15
    </span>
    <span
      style="width: 10%;"
    >
      10
    </span>
    <span
      style="width: 10%;"
    >
      <button
        class="button buttonSmall"
        type="button"
      >
        <svg
          height="18px"
          width="18px"
        >
          check.svg
        </svg>
      </button>
    </span>
  </div>
</div>
`;

exports[`Search Component  Snapshot test 1`] = `
<form
  class="searchForm"
>
  <label
    class="label"
    for="search"
  >
    Search:
  </label>
   
  <input
    class="input"
    id="search"
    type="text"
    value="React"
  />
  <p>
    Searching : 
    <strong>
      React
    </strong>
  </p>
  <button
    class="button buttonSmall"
    type="submit"
  >
    Submit
  </button>
</form>
`;
